### ISSUE TITLE
WebTorrent Memory Leak

### ISSUE THREAD

**Issue by Ryan6578** [webtorrent/webtorrent#2356]
**What version of this package are you using?**
WebTorrent 1.8.26

**What operating system, Node.js, and npm version?**
Ubuntu 22.04 Server, Node.JS 18.7.0, NPM 8.15.0

**What happened?**
I created a Discord bot to allow friends to torrent content to my Plex server through the bot without having to ask me. I noticed that memory usage would creep up until the process was eventually killed by systemd for being too swappy. I worked through debugging my code and couldn't find any obvious leaks, so I made a test app and confirmed that the leak was happening in WebTorrent. Test code is as follows:

```
const WebTorrent = require('webtorrent');

const client = new WebTorrent();

const magnets = []; //magnet links here - torrents them one at a time after the previous one has finished

client.add(magnets.pop(), { path: './', destroyStoreOnDestroy: false }, (torrent) => { updateProgress(torrent); });

function updateProgress(torrent) {
    // Record whether or not the torrent has completed
    const done = torrent.done;

    console.log(`Downloading '${torrent.name}': ${Math.round(torrent.progress * 100)}%`);

    if(done) {
      torrent.destroy();
      if(magnets.length > 0)
      client.add(magnets.pop(), { path: './', destroyStoreOnDestroy: false }, (torrent) => { updateProgress(torrent); });
    } else {
      setTimeout(() => { updateProgress(torrent); }, 3000);
    }
  }
```

I took a heapdump of the process running with the Discord bot, and most of the memory is being utilized by `system/JSArrayBufferData`, specificially in Connection/UTP/Wire. Let me know if you'd like me to post a snapshot of the heapdump.

**What did you expect to happen?**
Chrome's V8 should garbage collect objects after a torrent has completed.

**Are you willing to submit a pull request to fix this bug?**
If I could be pointed to where the issue lies (whether it be the store, in WebTorrent, etc).

**Comment by Ryan6578** [comment]
Can anyone confirm they're seeing this on their end as well?

**Comment by vankasteelj** [comment]
Yes, here: https://github.com/webtorrent/webtorrent/issues/1026

**Comment by Abdulla060** [comment]
I have the exact same problem. I'm using Webtorrent on my raspberry pi and it keeps crashing due to low memory.

**Comment by ThaUnknown** [comment]
this is likely because your storage write speeds are lower than your download speeds, I recommend u reduce the DL speed

**Comment by Abdulla060** [comment]
@ThaUnknown just tried reducing the download speed to 10% of the maximum. still, the same problem although it happens much later

**Comment by ThaUnknown** [comment]
> @ThaUnknown just tried reducing the download speed to 10% of the maximum. still, the same problem although it happens much later

10% isnt very specific
well 10% of say 1Gbit is 100Mbit so 12MB/s, with a poor controller it still might be too much, and it likely is by the self evident "it happens much later" meaning that its likely still too high

**Comment by Abdulla060** [comment]
@ThaUnknown sorry i wasn't thinking when i wrote the comment. I reduced the download speed to 1MB/s. My rpi write speed is around 30MB/s.

**Comment by ThaUnknown** [comment]
@Abdulla060 
okay, that's more constructive, let's try to continue to narrow this down so we can find the potential underlying issue, do you have any custom settings defined in the object constructor?

**Comment by Abdulla060** [comment]
@ThaUnknown only the `downloadLimit` which is currently set to 1 * 1024 *1024 and nothing else.

**Comment by ThaUnknown** [comment]
any other code running? how many torrents are you running? what are their sizes?

**Comment by Abdulla060** [comment]
I only have the NodeJS code running. And im downloading 6 torrents each is about 350 MB simultaneously

**Comment by ThaUnknown** [comment]
how much memory are you working with?

**Comment by Abdulla060** [comment]
2 GB of memory with 1GB swap file

**Comment by ThaUnknown** [comment]
welp, I'm out of ideas, webtorrent rarely should use more than 100-200MB of RAM, you could try reducing that by setting `storeCacheSlots` to 0 in [client.add](https://github.com/webtorrent/webtorrent/blob/master/docs/api.md#clientaddtorrentid-opts-function-ontorrent-torrent-) opts and lowering `maxConns` < 55 in the [client constructor ](https://github.com/webtorrent/webtorrent/blob/master/docs/api.md#client--new-webtorrentopts)

**Comment by Abdulla060** [comment]
sadly, no change. I'll try to debug it over the weekend. for now, my project is on halt.

**Comment by ThaUnknown** [comment]
> sadly, no change. I'll try to debug it over the weekend. for now, my project is on halt.

you could try finding where all the memory is allocated if you know your way around the node debugger, if you manage to do that I'll probs fix it same day

**Comment by risharde** [comment]
I want to confirm that there is definitely a really bad memory leak. On a 1 GB VPS, after about 4 torrent downloads, memory is exhausted, application in NodeJS crashes. Very difficult to provide any traces unfortunately :(

**Comment by risharde** [comment]
Forcing a gabage collection using the --expose-gc and running the appropriate code does not help eventually either.

**Comment by ThaUnknown** [comment]
@Abdulla060 this should be mostly mitigated now

**Comment by Ryan6578** [comment]
> @Abdulla060 this should be mostly mitigated now

Specifically @Abdulla060's comment? Or the WebTorrent memory leak in general?

**Comment by WZY1999428** [comment]
Hello, when I use a popular download resource, the memory usage will increase continuously when it reaches 2g. This is the link magnet:? xt=urn:btih:98901dd089140a8752e0e0e18e75211eee3baf20

**Comment by vankasteelj** [comment]
It still is relevant

**Comment by Ryan6578** [comment]
> It still is relevant

Maybe it's just me, but with the latest release + latest NodeJS, I haven't noticed any memory leaks.

**Comment by vankasteelj** [comment]
Oh I cannot test locally RN since the ESM switch for which I'm not ready yet. But there has been no commit referencing or touching the issue, it's juste the stale-bot that decided to come here. If it's resolved, great, but that has to be a manual closure, we can't let that bot stale this as it (afaik) hasnt been adressed yet

**Comment by ThaUnknown** [comment]
there's no commit referencing the issue because the issue is vauge, I fixed 2 memory leaks, but I'm not sure if they were related to this issue, which is why I honestly wanted to close this because reports like "high cpu usage" or "high ram usage" are useless without a reason of tracing of what's actually wrong

**Comment by vankasteelj** [comment]
Idk @ThaUnknown 

I've been unable to update since 0.108 because of this very issue. I've tried and reported multiple time the problem across multiple versions, all of which were closed and sent back here to centralize the issue. I haven't been testing every update because nobody ever seem to reference this issue or another one relating to the poor memory/CPU management plaguing Webtorrent since >0.108. So I don't see why it would be solved. I've even backported the speed limiter option to 0.108 because I cannot use a newer version but still like the idea of the news features that are implemented.

And now you have moved on to ESM, great but I'm not ready for it. That's technology debt on my own, but especially if I need to rework my entire codebase in order to find out that maybe the very same issue we're reporting for years still is a problem, it's tons of work.

So my only point was that the stale bot shouldn't be the one to decide if it's solved or not. If it's fixed it should be closed. If fixes have been made referencing this issue, there should be comments in commits referencing this issue. Idk I'm not a maintainer here, just someone that hopes that this blocking issue (for me and others) will be fixed at some point. No need to judge because we don't move to ESM as fast as others...

**Comment by ThaUnknown** [comment]
the stale bot closes issues which aren't accepted, and this issue isn't accepted because it lacks any examples, information or detail, so it will get closed if noone makes a valid report or if its not active enough

**Comment by vankasteelj** [comment]
Yes, maybe centralizing in https://github.com/webtorrent/webtorrent/issues/1026 - which is accepted and seems to be similar/related would be a solution to close this one. Idk I've commented multiple issues but have no track of which is which, just GitHub notifications

**Comment by ThaUnknown** [comment]
that's regarding CPU managment, not memory, which I've been addressing over a lot of PR's

**Comment by ThaUnknown** [comment]
> All I can say is that every pre-esm version had the same issue.

even 1.9.x? Those migrated away from stream and buffer to uint8, iterators and streamx, which for many operations reduced CPU load by up to 40%, tho where the IO for them was the biggest [p2p] it still uses stream ;-;

**Comment by vankasteelj** [comment]
I believe so but I can test that some day if you want to be sure

**Comment by vankasteelj** [comment]
@ThaUnknown  I've tested 1.9.7 and it does seem to work perfectly. Maybe I stopped testing new versions at 1.8.x or something. 

On a secondary note but we've been talking about it: I've also looked at ESM, and remembered why I cant upgrade webtorrent > 2.x: I'm using NWJS, which like Electron, doesnt accept ESM at this time. I guess it's the same issue with Webtorrent-desktop being stuck at 1.9.7, pre-esm version.
